### DISCUSSION TITLE
Allow 'ref readonly struct' for 'readonly ref struct'

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8752] [Language Ideas]
@mazhou-msft commented on [Thu Mar 01 2018](https://github.com/dotnet/roslyn/issues/25187)

**Version Used**: 
Visual Studio 2017 27428.0.d15.7

**Steps to Reproduce**:

1. Create a C# project for netcoreapp2.1
2. Write a declaration of a ref struct:
```C#
public readonly ref struct S { }
```

**Expected Behavior**:
The order of 'ref' and 'readonly' should not matter; I can either use 'readonly ref struct' or 'ref readonly struct' to declare readonly ref-like types.

**Actual Behavior**:
'ref readonly struct' is not allowed. looks like `ref struct` is a compound keyword like `partial class`.

---

@CyrusNajmabadi commented on [Thu Mar 01 2018](https://github.com/dotnet/roslyn/issues/25187#issuecomment-369824078)

This is not a compiler issue, but a request for a change in the language.  as such, this should be filed over at dotnet/csharplang.  Thanks!

---

@alrz commented on [Thu Mar 01 2018](https://github.com/dotnet/roslyn/issues/25187#issuecomment-369830591)

this is already championed (https://github.com/dotnet/csharplang/issues/946)

---

@jcouv commented on [Fri Mar 02 2018](https://github.com/dotnet/roslyn/issues/25187#issuecomment-370113464)

Thanks for reporting this. 
Since this is a known issue, I'll go ahead and close as duplicate. @alrz is close to getting this fix (https://github.com/dotnet/roslyn/pull/23533) into C# 7.3.

---

@gafter commented on [Thu May 17 2018](https://github.com/dotnet/roslyn/issues/25187#issuecomment-389934625)

This was not championed or fixed. What was fixed was the ordering of `ref` and `partial`. There is still an issue with the ordering of `ref` and `readonly`.

**Comment by bondsbw** [comment]
#1530

**Comment by stakx** [comment]
I'm happy to close my issue (#1530) in favour of this one although it's perhaps the more general of the two. Has it already been decided that relaxation should occur only with `struct` declarations (instead of return value type declarations)?
